packages/grid/x-data-grid/src/components/GridPagination.tsx

```diff
@@ -29,60 +31,70 @@ export const GridPagination = React.forwardRef<HTMLDivElement, Partial<TablePagi
     const apiRef = useGridApiContext();
     const rootProps = useGridRootProps();
     const paginationState = useGridSelector(apiRef, gridPaginationSelector);
+    const visibleTopLevelRowCount = useGridSelector(apiRef, gridFilteredTopLevelRowCountSelector);
+
+    const rowCount = React.useMemo(
+      () => rootProps.rowCount ?? visibleTopLevelRowCount ?? 0,
+      [rootProps.rowCount, visibleTopLevelRowCount],
+    );
 
     const lastPage = React.useMemo(
-      () => Math.floor(paginationState.rowCount / (paginationState.pageSize || 1)),
-      [paginationState.rowCount, paginationState.pageSize],
+      () => Math.floor(rowCount / (paginationState.paginationModel.pageSize || 1)),
+      [rowCount, paginationState.paginationModel.pageSize],
     );
 
     const handlePageSizeChange = React.useCallback(
       (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
-        const newPageSize = Number(event.target.value);
-        apiRef.current.setPageSize(newPageSize);
+        const pageSize = Number(event.target.value);
+        apiRef.current.setPageSize(pageSize);
       },
       [apiRef],
     );
 
     const handlePageChange = React.useCallback<TablePaginationProps['onPageChange']>(
-      (event, page) => {
+      (_, page) => {
         apiRef.current.setPage(page);
       },
       [apiRef],
     );
 
     if (process.env.NODE_ENV !== 'production') {
       // eslint-disable-next-line react-hooks/rules-of-hooks
-      const warnedOnceMissingPageSizeInRowsPerPageOptions = React.useRef(false);
+      const warnedOnceMissingInPageSizeOptions = React.useRef(false);
+      const pageSize =
+        rootProps.paginationModel?.pageSize ?? paginationState.paginationModel.pageSize;
       if (
-        !warnedOnceMissingPageSizeInRowsPerPageOptions.current &&
+        !warnedOnceMissingInPageSizeOptions.current &&
         !rootProps.autoPageSize &&
-        !rootProps.rowsPerPageOptions.includes(rootProps.pageSize ?? paginationState.pageSize)
+        !rootProps.pageSizeOptions.includes(pageSize)
       ) {
         console.warn(
           [
-            `MUI: The page size \`${
-              rootProps.pageSize ?? paginationState.pageSize
-            }\` is not preset in the \`rowsPerPageOptions\``,
+            `MUI: The page size \`${pageSize}\` is not preset in the \`pageSizeOptions\``,
             `Add it to show the pagination select.`,
           ].join('\n'),
         );
 
-        warnedOnceMissingPageSizeInRowsPerPageOptions.current = true;
+        warnedOnceMissingInPageSizeOptions.current = true;
       }
     }
 
     return (
       <GridPaginationRoot
         ref={ref}
         component="div"
-        count={paginationState.rowCount}
-        page={paginationState.page <= lastPage ? paginationState.page : lastPage}
+        count={rowCount}
+        page={
+          paginationState.paginationModel.page <= lastPage
```

Comment:
Maybe extract `paginationState.paginationModel` to a variable to reduce the chaining a bit and make the code look more compact?